sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogAndTables.scala

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.{AlterTable, DeleteFromTable, DescribeTable, LogicalPlan, SubqueryAlias}
+import org.apache.spark.sql.catalyst.plans.logical.sql.{AlterTableAddColumnsStatement, AlterTableAlterColumnStatement, AlterTableDropColumnsStatement, AlterTableRenameColumnStatement, AlterTableSetLocationStatement, AlterTableSetPropertiesStatement, AlterTableUnsetPropertiesStatement, AlterViewSetPropertiesStatement, AlterViewUnsetPropertiesStatement, DeleteFromStatement, DescribeColumnStatement, DescribeTableStatement, UpdateTableStatement}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.{CatalogManager, LookupCatalog, Table, TableChange, V1Table}
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+
+/**
+ * Resolves catalogs and tables from the multi-part identifiers in SQL statements, and convert the
+ * statements to the corresponding v2 commands if the resolved table is not a [[V1Table]].
+ */
+class ResolveCatalogAndTables(val catalogManager: CatalogManager)
```

Comment:
This doesn't keep catalog and table resolution separate, it combines them in the `CatalogAndTable` extractor.

I think those should be separate rules. Resolving the catalog should create an `UnresolvedRelation`, then resolving the table should replace it with a DSv2 relation.

This approach also keeps the connector API separate from logical plans. There's no need to make the logical `AlterTable` node dependent on the `Table` API because it is a purely SQL operation and it should not be tied to the underlying connector API.

Reply:
I tried your idea, but unfortunately it doesn't work. For example, when resolving ALTER TABLE, what we need to do is:
1. resolve catalog
2. resolve table
3 if table is v1, convert to v1 ALTER TABLE command. Otherwise, convert to v2 ALTER TABLE command.

These 3 steps must be done together, otherwise we need to introduce a new logical plan as an intermediate representation of the query plan that has catalog resolved, and will be converted to v1 or v2 command later.

Reply:
These steps must be done together only for v1 tables, and we already have rules that convert v1 tables separately. So the v1 table rules will need to look up the table before converting. I agree with you there. But when the catalog is any v2 catalog other than the session catalog, it will work to use UnresolvedRelation.